src/compiler/parser.ts

```diff
@@ -1888,7 +1888,12 @@ namespace ts {
             return false;
         }
 
-        const viableKeywordSuggestions = Object.keys(textToKeywordObj).filter(keyword => keyword.length > 2);
+        let viableKeywordSuggestions: () => string[] = () => {
+            const result = Object.keys(textToKeywordObj()).filter(keyword => keyword.length > 2);
+            // Reassign the function, so we do not have an if statement in this lazy evaluation.
+            viableKeywordSuggestions = () => result;

```

Comment:
FWIW we have a helper named `memoize` for this sort of thing; I'm not sure if this method actually turns out to be faster (I did some benchmarking and it seemed to be a wash).

Reply:
Ah, good to know. I also benchmarked `memoize` and expected it to be slower due to the additional branch, but it was almost identical, so I guess v8 JIT optimises the empty branch of `memoize` away.

Reply:
After further testing, it seems rather the opposite. My code above gets deoptimised. An alternative version using a object is almost 7x faster and is fully optimised by v8, running at the max possible speed (looks like the function gets inlined)

```javascript
1,005,352,943.44 ops/sec ±0.13% 0.0000000009946756 sec/op baseline
  153,337,715.31 ops/sec ±0.15% 0.0000000065215528 sec/op memoize1
1,004,871,904.32 ops/sec ±0.17% 0.0000000009951517 sec/op memoize2
```

memoize1 is the original implementation in tsc, memoize2 is the new one.

```typescript
    function a() {
        return [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];
    }

    const array = a();
    bench.add('baseline', () => {
        array[2];
    });

    function memoize<T>(callback: () => T): () => T {
        let value: T;
        return () => {
            if (callback) {
                value = callback();
                callback = undefined!;
            }
            return value;
        };
    }

    const mem1 = memoize(a);
    bench.add('memoize1', () => {
        mem1()[2];
    });


    function memoize2<T>(callback: () => T): {get(): T}  {
        const cache = {
            get() {
                const result = callback();
                cache.get = function(){ return result; };
                return result;
            }
        }
        return cache;
    }

    const mem2 = memoize2(a);

    bench.add('memoize2', () => {
        mem2.get()[2];
    });

```

The good thing is we can remove the ugly function reassignment of the function from the user-side to the memoize implementation, disadvantage is that the calling signature slightly changes.

Reply:
Actually, returning an arrow function that simply redirects the call allows the v8 to optimise it, too. This keeps the call signature intact.

```typescript
    function memoize2<T>(callback: () => T): () => T {
        const cache = {
            get() {
                const result = callback();
                cache.get = function(){ return result; };
                return result;
            }
        }
        return () => cache.get();
    }
```

I pushed a new commit with that change. Would be interesting to see if it has any perf impact since memoize is used in a few places and an almost 7x improvement could have a slight impact.

Reply:
I am curious about that new version, yeah. If you send a PR with it, I could rerun the benchmark and see.

Reply:
@jakebailey it's already part of this PR, so if you can run benchmarks on this closed PR, we will see it soon :)

Reply:
Not sure it'll work, since it tests the merge head. I'll just reopen and see.